### ISSUE TITLE
Add Ceph RBD storage driver

### ISSUE THREAD

**Issue by cduchesne** [codedellemc/rexray#527]
emccode/libstorage#191

v1
======
- [x] RBD executor
- [x] ls volumes
- [x] local device mapping
- [x] attachment filters
- [x] volume status
- [x] volume inspect
- [x] volume create
- [x] multi-pool support
- [x] volume delete
- [x] volume attach
- [x] volume mount
- [x] volume unmount
- [x] volume detach
- [x] remote attachment status inferred from RBD watchers
- [x] tests
- [x] docs
- [x] Add vagrantfile for testing

Progress here: https://github.com/codenrhoden/libstorage/tree/feature/rbd

**Comment by codenrhoden** [comment]
I got Volume DELETE working today, but ran into a problem we will have to choose a direction on...

I've wanted to use the Ceph "image spec" for a full volumeID, which takes the form of `[pool/]<name>[@snap]`.  That way I can do things like:

```
rexray volume get mypool/image1
rexray volume create rbd/image --size 10
rexray volume delete image
```

In the above example, anything before `/` represents the pool, and the driver knows how to deal with it. But if there isn't a pool given, `rbd` is used as the default pool (and this can be overriden via config).

however, using these formats requires the volumeIDs to be url-encoded, and our HTTP router has some issues with that.  See: https://github.com/gorilla/mux/issues/132.  This was supposed fixed with https://github.com/gorilla/mux/commit/0a192a193177452756c362c20087ddafcf6829c4, but my testing didn't work when I set that option. I had to fall back to the `{name:.*}` matching regex, which _completely works_ in our setup, but isn't quite as future prof if we ever wanted to have a URL like `.../{service}/{volumeID}/{anything}`.  All of our routes end at `/{volumeID` so we'd be okay for now, but I'd rather the `UseEncodedPath()` setting just did the right thing.

Will try to debug that next week. It should work!

**Comment by akutz** [comment]
Why don't you use a custom query string. Those get thrown into the store.

-- 
-a

**Comment by codenrhoden** [comment]
@akutz: Thanks for the pointer. I thought about it over the weekend, and it seems that your suggestion would require custom parsing within rexray to deal with that and pass the extra values for the query string down to libstorage. put another way, I'm not sure how I'd do it without making it driver specific. I could be very wrong there, but...

@clintonskitson also pointed out to me on Slack that other drivers are using `VolumeType` as a holder for pool information. Indeed I found this: https://github.com/codedellemc/libstorage/blob/master/drivers/storage/scaleio/storage/scaleio_storage.go#L674-L679

@akutz, unless I'm way off base, using `VolumeType` seems the more consistent way to go.

**Comment by akutz** [comment]
Hi @codenrhoden,

That definitely seems to be a better solution. Thanks for finding that!

**Comment by codenrhoden** [comment]
Just an update here...  Only thing holding up generic end-to-end testing is Volume Mount. That's not working quite yet, and is depending on changes from this thread: https://github.com/codedellemc/rexray/pull/619.  This intertwined with attachment filters and volume status. All three should be taken care of at once.

I'm closely tracking Libstorage 0.3.3 release candidates, and rexray master.

After mounts are in place, that leaves high-order funcionality like snapshots and volume copies.

Adding RBD advisory locks is a big deal, and I'd like to tackle that after an end-to-end testable piece is built for early testing. And speaking of testing, I need to add a lot of them to the source. :) While I'm coming up to speed on golang in general, golang testing is new to me still. May have to get a tutorial on that. :) Thankfully there are lots of examples to look at.

**Comment by codenrhoden** [comment]
Another update here...

After a bit of testing, and keeping up with REX-Ray 0.6.1 and LS 0.3.3 upcoming releases, I think I have the minimum for something useful. The current state of the driver will be slow when there are lots of volumes, because I have no choice but to make several API calls for each volume in order to get their size and infer their attachment status. Right now this all happens in serial.

There are two ways to speed it up -- first is to make all the needed calls in goroutines so that they are done in parallel (with some limiter on how many to do in parallel). But the better way to do it is to enhance `go-ceph` to implement the async routines that are in the C API and implement smarter methods in `go-ceph` to just get that data in parallel and asynchronously for me.  Future enhancements!

From what I've gathered from @akutz, a plan would be to submit what I have as a PR, ultimately destined for Libtorage 0.3.4. To facilitate that, I need to add my tests, a testing doc/plan, and add a Vagrantfile to get people up and testing easily.

**Comment by a-nldisr** [comment]
@codenrhoden 
At what amount of volumes are you running into issues?

**Comment by codenrhoden** [comment]
@a-nldisr good question. I don't have the hard data to back anything up. Ultimately I do need to write a Go benchmark to show the difference. Subjectively, I can say that when running on VMs (a Vagrant environment running on virtualbox), that I can see the difference between 1 RBD and 10 RBDs -- so not very many at all. The difference here being maybe 1 second to get a response, and 3-4 seconds to get a response.

For what it's worth, in the short term I am abandoning the use of `go-ceph`, so this will no longer be an issue. The reason for this is two-fold:

(1) `go-ceph` does not have APIs for volume attach/detach (e.g. `rbd map` and `rbd unmap`. I have to call out to the `rbd` executable to do that no matter what, so `go-ceph` utility is already limited.
(2) Some of the methods implemented in `go-ceph` are woefully behind the implementations in the `rbd` executable. The prime example of this is `rbd rm`. In the `rbd` executable, it was enhanced many releases ago to be use the async API, and to delete many objects in parallel (i think default is 20). `go-ceph` is using the sync API, and is deleting objects in serial. I had forgotten this used to ever be the case, because this problem went away so long ago. This can be difference between taking a few seconds to delete an image, and several minutes.

Long term I would still like to improve `go-ceph`, but it is just that, a long-term goal.

**Comment by a-nldisr** [comment]
In that case, perhaps this code helps: 
https://github.com/medallia/docker/blob/ca6c07f0a093b9f8401059fe55946b8076bf0fea/volume/ceph/ceph.go

Will keep an eye on this, perhaps we can help with some testing once its nearing completion.

**Comment by codenrhoden** [comment]
@a-nldisr there is a PR open for early review. You can check it out here: https://github.com/codedellemc/libstorage/pull/347

**Comment by codenrhoden** [comment]
Tests have been added to the.

Vagrant file will be based off the work you see here: https://github.com/codedellemc/vagrant/pull/30

Just need to add docs now. Getting close!

**Comment by codenrhoden** [comment]
@cduchesne I'd like to take the *v2* and *v3* functionality in this PR and move them into their own issues. Or perhaps break them up even more...  That way we can close this PR when rex 0.7.0 is release and we can more strategically target new features to upcoming releases.  in fact, most of the added functionality in here probably belongs in libstorage issues rather than rexray.  Thoughts?

**Comment by tobilg** [comment]
@codenrhoden Great to see the progress! A quick question: Do you plan to inlcude all of v1/v2/v3 in the final v0.7.0 release?

**Comment by a-nldisr** [comment]
I tried to cross compile this on a Macbook to build artifacts for linux i can test on a system. But i ran into the follow command that suddenly switches and then outputs binaries for darwin:

**Comment by a-nldisr** [comment]
Can you tell me what is meant with volume locking?

**Comment by codenrhoden** [comment]
@tobilg The functionality in 0.7.0 will only cover v1. I am intending to break out all the future enhancements into separate issues, as long as I can get @cduchesne to sign off on that. See https://github.com/codedellemc/rexray/issues/527#issuecomment-267719952

**Comment by codenrhoden** [comment]
@a-nldisr By volume locking, I mean that after an `rbd map` is performed, an additional `rbd lock add` will be performed to signify that the image is in-use. Any future volume attaches (e.g. `rbd map`'s) will first check to see if there is a lock on the RBD before mapping.

This is more reliable than the current method of looking for RBD watchers to see if the volume is already in use.

**Comment by tobilg** [comment]
@cduchesne Thanks for the feedback! V1 would already be great... When will the final 0.7.0 arrive approximately?

**Comment by codenrhoden** [comment]
I've opened a slew of issues in libstorage, all with the rbd label, to capture future work for this driver.  See: https://github.com/codedellemc/libstorage/issues?q=is%3Aopen+is%3Aissue+label%3Arbd

**Comment by codenrhoden** [comment]
@tobilg We are in late-stage release candidate testing right now, so my hope would be within the next week!

**Comment by a-nldisr** [comment]
@codenrhoden Thanks! So in the scenario of a top of rack switch failure where jobs continue to run inside the affected rack and have a rbd device and then are rescheduled somewhere else by the scheduler, you would be in serious trouble without this volume locking if the rack recovers am i right?
